### PR TITLE
feat: Implement Delegate Resignation (AIP11)

### PR DESCRIPTION
**Pull request by vasild**

### CODE REVIEW

File: packages/core-jest-matchers/src/transactions/types/delegate.ts [hunk 1]
```diff
@@ -17,7 +17,8 @@ expect.extend({
     toBeDelegateType: received => {
         return {
             message: () => "Expected value to be a valid DELEGATE transaction.",
-            pass: received.type === DelegateRegistration,
+            pass: received.type === TransactionTypes.DelegateRegistration ||
+                  received.type === TransactionTypes.DelegateResignation,
```
Comment (faustbrian) [inline packages/core-jest-matchers/src/transactions/types/delegate.ts]: This matcher is supposed to check for a delegate registration, this will now cause false results for tests if you don't pay attention. All changes in this file need to be reverted.

There is already a matcher for a resignation at https://github.com/ArkEcosystem/core/blob/b58f7dcfca73a34158f7d76486bcf3e11dcfa596/packages/core-jest-matchers/src/transactions/types/delegate-resignation.ts.

File: packages/crypto/src/transactions/types/schemas.ts [hunk 2]
```diff
@@ -215,9 +215,23 @@ export const multiPayment = extend(transactionBaseSchema, {
 
 export const delegateResignation = extend(transactionBaseSchema, {
     $id: "delegateResignation",
+    required: ["asset"],
     properties: {
         type: { transactionType: TransactionTypes.DelegateResignation },
         amount: { bignumber: { minimum: 0, maximum: 0 } },
+        asset: {
```
Comment (faustbrian) [inline packages/crypto/src/transactions/types/schemas.ts]: Why do you need an asset? The transaction will be applied to the sender which is required to be a wallet with a username so you can grab the username from it directly.

File: packages/core-state/src/wallets/wallet-manager.ts [hunk 3]
```diff
@@ -129,19 +129,16 @@ export class WalletManager implements State.IWalletManager {
     }
 
     public loadActiveDelegateList(roundInfo: Shared.IRoundInfo): State.IDelegateWallet[] {
+        const delegates: State.IWallet[] = this.buildDelegateRanking(undefined, roundInfo);
```
Comment (supaiku0) [inline packages/core-state/src/wallets/wallet-manager.ts]: Why this change? Line 325 should be enough.

**Comment by vasild** [comment]
Alright, I added a `resigned` flag to mark delegates that have resigned instead of un-delegating them.

> The sender must be a delegate

The check is here: https://github.com/ArkEcosystem/core/pull/2538/files#diff-304db985d999842253be956ceec47a2aR20 `!wallet.username`

> Once forged the delegate is irreversibly deactivated

Will be handled by existing code: https://github.com/ArkEcosystem/core/blob/afb69f2141383/packages/core-transactions/src/handlers/delegate-registration.ts#L60

> no longer possible to vote for this delegate

Added a check here: https://github.com/ArkEcosystem/core/pull/2538/files#diff-ced27755786a5f642370021ba08a2853R66

> no longer treated as an active delegates, regardless of vote balance

Removed resigned delegates: https://github.com/ArkEcosystem/core/pull/2538/files#diff-097e9ec56428de146b54cea1f67d0448R325

File: packages/core-state/src/wallets/wallet.ts [hunk 4]
```diff
@@ -164,6 +166,10 @@ export class Wallet implements State.IWallet {
             audit.push({ "New username": username });
         }
 
+        if (transaction.type === Enums.TransactionTypes.DelegateResignation) {
+            audit.push({ "Resigned delegate": this.username });
```
Comment (supaiku0) [inline packages/core-state/src/wallets/wallet.ts]: Don't need an asset.

File: packages/core-transactions/src/errors.ts [hunk 5]
```diff
@@ -90,6 +90,15 @@ export class WalletUsernameAlreadyRegisteredError extends TransactionError {
     }
 }
 
+export class WalletUsernameMismatchError extends TransactionError {
```
Comment (supaiku0) [inline packages/core-transactions/src/errors.ts]: The resignation can only be made by delegates. Therefore we don't need a username.

File: packages/core-transactions/src/handlers/delegate-resignation.ts [hunk 6]
```diff
@@ -8,21 +9,59 @@ export class DelegateResignationTransactionHandler extends TransactionHandler {
     }
 
     public async bootstrap(connection: Database.IConnection, walletManager: State.IWalletManager): Promise<void> {
-        return;
+        // XXX should we do something here?
```
Comment (supaiku0) [inline packages/core-transactions/src/handlers/delegate-resignation.ts]: Yes, here we have to update the wallets and apply the `resignated`  status.
See the other handlers, but basically:
1) Get all resignation assets
2) Iterate over them and apply resignated flag to sender wallet

File: packages/core-transactions/src/handlers/delegate-resignation.ts [hunk 7]
```diff
@@ -8,21 +9,59 @@ export class DelegateResignationTransactionHandler extends TransactionHandler {
...
     }
 
     public canBeApplied(
         transaction: Interfaces.ITransaction,
         wallet: State.IWallet,
         databaseWalletManager: State.IWalletManager,
     ): boolean {
+        if (!wallet.username || !transaction.data.asset.delegate.username) {
```
Comment (supaiku0) [inline packages/core-transactions/src/handlers/delegate-resignation.ts]: The transaction doesn't need an asset.

File: packages/core-transactions/src/handlers/delegate-resignation.ts [hunk 8]
```diff
@@ -8,21 +9,59 @@ export class DelegateResignationTransactionHandler extends TransactionHandler {
...
+            throw new WalletUsernameEmptyError();
+        }
+
+        if (wallet.username !== transaction.data.asset.delegate.username) {
```
Comment (supaiku0) [inline packages/core-transactions/src/handlers/delegate-resignation.ts]: Can be removed.

File: packages/crypto/src/transactions/builders/transactions/delegate-resignation.ts [hunk 9]
```diff
@@ -11,12 +11,20 @@ export class DelegateResignationBuilder extends TransactionBuilder<DelegateResig
         this.data.type = TransactionTypes.DelegateResignation;
         this.data.fee = feeManager.get(TransactionTypes.DelegateResignation);
         this.data.amount = BigNumber.ZERO;
-        this.data.asset = {};
+        this.data.recipientId = undefined;
+        this.data.senderPublicKey = undefined;
+        this.data.asset = { delegate: {} } as ITransactionAsset;
+    }
+
+    public usernameAsset(username: string): DelegateResignationBuilder {
```
Comment (supaiku0) [inline packages/crypto/src/transactions/builders/transactions/delegate-resignation.ts]: Remove.

File: packages/crypto/src/transactions/builders/transactions/delegate-resignation.ts [hunk 10]
```diff
@@ -11,12 +11,20 @@ export class DelegateResignationBuilder extends TransactionBuilder<DelegateResig
...
+        this.data.asset.delegate.username = username;
+        return this;
     }
 
     public getStruct(): ITransactionData {
         const struct: ITransactionData = super.getStruct();
         struct.amount = this.data.amount;
+        struct.recipientId = this.data.recipientId;
```
Comment (supaiku0) [inline packages/crypto/src/transactions/builders/transactions/delegate-resignation.ts]: Remove.

File: packages/crypto/src/transactions/serializer.ts [hunk 11]
```diff
@@ -69,7 +69,8 @@ export class Serializer {
                 break;
             }
 
-            case TransactionTypes.DelegateRegistration: {
+            case TransactionTypes.DelegateRegistration:
+            case TransactionTypes.DelegateResignation: {
```
Comment (supaiku0) [inline packages/crypto/src/transactions/serializer.ts]: Revert. This is legacy V1 serialization.

File: packages/crypto/src/transactions/types/delegate-resignation.ts [hunk 12]
```diff
@@ -12,10 +12,22 @@ export class DelegateResignationTransaction extends Transaction {
     }
 
     public serialize(options?: ISerializeOptions): ByteBuffer {
-        return new ByteBuffer(0);
```
Comment (supaiku0) [inline packages/crypto/src/transactions/types/delegate-resignation.ts]: Revert.

File: packages/crypto/src/transactions/types/delegate-resignation.ts [hunk 13]
```diff
@@ -12,10 +12,22 @@ export class DelegateResignationTransaction extends Transaction {
...
+        const buffer: ByteBuffer = new ByteBuffer(delegateBytes.length, true);
+
+        buffer.writeByte(delegateBytes.length);
+        buffer.append(delegateBytes, "hex");
+
+        return buffer;
     }
 
     public deserialize(buf: ByteBuffer): void {
-        return;
```
Comment (supaiku0) [inline packages/crypto/src/transactions/types/delegate-resignation.ts]: Revert.

File: packages/core-state/src/wallets/wallet-manager.ts (hunk 3 above)
Comment (vasild) [reply packages/core-state/src/wallets/wallet-manager.ts]: This change is because we want to check against `maxDelegates` with resigned delegates removed. Thus need to call `buildDelegateRanking()` before the check because `allByUsername()` would not filter anything.

Also this change makes the code simpler because it eliminates the call to `allByUsername()` and removes the variable `delegatesWallets`, so one call and one variable less.

Btw what a wretched naming for `maxDelegates`! We actually want the number of delegates to be bigger than that and emit an error if it is *less*. `minDelegates` would be more appropriate, but out of the scope of this PR.

File: packages/core-transactions/src/handlers/delegate-resignation.ts [hunk 14]
```diff
@@ -8,21 +9,58 @@ export class DelegateResignationTransactionHandler extends TransactionHandler {
...
+            walletManager.findByPublicKey(transaction.senderPublicKey).resigned = true;
+        }
     }
 
     public canBeApplied(
         transaction: Interfaces.ITransaction,
         wallet: State.IWallet,
         databaseWalletManager: State.IWalletManager,
     ): boolean {
+        if (!wallet.username) {
```
Comment (supaiku0) [inline packages/core-transactions/src/handlers/delegate-resignation.ts]: Also throw if the sender already `resigned`.

File: packages/core-transactions/src/handlers/delegate-resignation.ts [hunk 15]
```diff
@@ -8,21 +9,58 @@ export class DelegateResignationTransactionHandler extends TransactionHandler {
...
+    public canEnterTransactionPool(
+        data: Interfaces.ITransactionData,
+        pool: TransactionPool.IConnection,
+        processor: TransactionPool.IProcessor,
+    ): boolean {
+        if (this.typeFromSenderAlreadyInPool(data, pool, processor)) {
+            processor.pushError(
+                data,
+                "ERR_PENDING",
+                `Delegate resignation for "${data.asset.delegate.username}" already in the pool`,
```
Comment (supaiku0) [inline packages/core-transactions/src/handlers/delegate-resignation.ts]: There's no asset. You need to get the username from the sender wallet.

File: __tests__/functional/transaction-forging/delegate-resignation.test.ts [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+            await expect(initialFunds).toBeAccepted();
+            await support.snoozeForBlock(1);
+            await expect(initialFunds.id).toBeForged();
+
+            // Resign a delegate
+            const transactionsResign = TransactionFactory.delegateResignation()
+                .withPassphrase(passphrase)
+                .createOne();
+
+            await expect(transactionsResign.id).toBeRejected();
```
Comment (faustbrian) [inline __tests__/functional/transaction-forging/delegate-resignation.test.ts]:
```suggestion
            await expect(transactionsResign).toBeRejected();
```

File: __tests__/functional/transaction-forging/delegate-resignation.test.ts [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+            await expect(transactionsResign1).toBeAccepted();
+            await support.snoozeForBlock(1);
+            await expect(transactionsResign1.id).toBeForged();
+
+            // Resign a delegate
+            const transactionsResign2 = TransactionFactory.delegateResignation()
+                .withPassphrase(passphrase)
+                .createOne();
+
+            await expect(transactionsResign2.id).toBeRejected();
```
Comment (faustbrian) [inline __tests__/functional/transaction-forging/delegate-resignation.test.ts]:
```suggestion
            await expect(transactionsResign2).toBeRejected();
```

File: __tests__/functional/transaction-forging/delegate-resignation.test.ts [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+            const passphrase = generateMnemonic();
+
+            // Initial Funds
+            const initialFunds = TransactionFactory.transfer(Identities.Address.fromPassphrase(passphrase), 100 * 1e8)
+                .withPassphrase(genesisPassphrase)
+                .createOne();
+
+            await expect(initialFunds).toBeAccepted();
+            await support.snoozeForBlock(1);
+            await expect(initialFunds.id).toBeAccepted();
```
Comment (faustbrian) [inline __tests__/functional/transaction-forging/delegate-resignation.test.ts]:
```suggestion
            await expect(initialFunds.id).toBeForged();
```
